digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -542,31 +460,45 @@ class NodesAndJunctionsService(roadwayDAO: RoadwayDAO, roadwayPointDAO: RoadwayP
           }
         }
 
-        /*
-          * R:road
-          * L:project link
-          * 0:junction point
-        */
-        // passed by collections "junctions in Head/ToTail" need to be passed for every iteration
-        roadsToHead.foreach { roadAddress: BaseRoadAddress => handleRoadsToHead(projectLink, getJunctionsInHead(projectLink, roadsToHead, roadsFromHead), roadAddress) }
+        // handle junction points for each project links
+        if ((roadsToHead ++ roadsFromHead).nonEmpty) {
+          createJunctionPointIfNeeded(projectLink, pos = BeforeAfter.After, projectLink.startAddrMValue, roadsToHead, roadsFromHead)
+        }
 
-        roadsFromHead.foreach { roadAddress: BaseRoadAddress => handleRoadsFromHead(projectLink, getJunctionsInHead(projectLink, roadsToHead, roadsFromHead), getJunctionsInHead(projectLink, roadsToHead, roadsFromHead), roadAddress) }
+        if ((roadsToTail ++ roadsFromTail).nonEmpty) {
+          createJunctionPointIfNeeded(projectLink, pos = BeforeAfter.Before, projectLink.endAddrMValue, roadsToTail, roadsFromTail)
+        }
 
-        roadsToTail.foreach { roadAddress: BaseRoadAddress => handleRoadsToTail(projectLink, getJunctionsToTail(projectLink, roadsToTail, roadsFromTail), getJunctionsInHead(projectLink, roadsToHead, roadsFromHead), roadAddress) }
+        // handle junction points for other roads, connected to each project link
+        roadsToHead.foreach { roadAddress: BaseRoadAddress =>
+          createJunctionPointIfNeeded(roadAddress, pos = BeforeAfter.Before, roadAddress.endAddrMValue, roadsToHead, roadsFromTail,
+            if (RoadAddressFilters.connected(projectLink)(roadAddress)) {
+              junctionPointDAO.fetchByRoadwayPoint(projectLink.roadwayNumber, projectLink.startAddrMValue, BeforeAfter.After).map(_.junctionId).toSeq.distinct.headOption
+            } else Some(NewIdValue))
+        }
 
-        roadsFromTail.foreach { roadAddress: BaseRoadAddress => handleRoadsFromTail(projectLink, getJunctionsToTail(projectLink, roadsToTail, roadsFromTail), getJunctionsInHead(projectLink, roadsToHead, roadsFromHead), roadAddress) }
-      }
-    }
-  }
+        roadsFromHead.foreach { roadAddress: BaseRoadAddress =>
+          createJunctionPointIfNeeded(roadAddress, pos = BeforeAfter.After, roadAddress.startAddrMValue, roadsToHead, roadsFromTail,
+            if (RoadAddressFilters.connected(projectLink)(roadAddress)) {
+              junctionPointDAO.fetchByRoadwayPoint(projectLink.roadwayNumber, projectLink.startAddrMValue, BeforeAfter.After).map(_.junctionId).toSeq.distinct.headOption
+            } else Some(NewIdValue))
+        }
```

Comment:
Found some problems here.., and also, found another different combination of operation that this fix doesn't support/fix. Yet!